packages/modules/cache/src/utils/parser.ts

```diff
@@ -0,0 +1,279 @@
+        return typeName
+      }
+    }
+
+    return null
+  }
+
+  /**
+   * Check if object structure matches GraphQL type fields
+   */
+  private objectMatchesType(obj: any, type: GraphQLObjectType): boolean {
+    const fields = type.getFields()
+    const objKeys = Object.keys(obj)
+
+    // Must have id field for entities
+    if (!obj.id || !fields.id) {
+      return false
+    }
+
+    // Check if at least 50% of non-null object fields match type fields
+    const matchingFields = objKeys.filter((key) => fields[key]).length
+    return matchingFields >= Math.max(1, objKeys.length * 0.5)
+  }
+
+  /**
+   * Get the expected type for a relationship field
+   */
+  private getRelationshipType(
+    parentType: string | null,
+    fieldName: string
+  ): string | undefined {
+    if (!parentType || !this.typeMap.has(parentType)) {
+      return undefined
+    }
+
+    const type = this.typeMap.get(parentType)!
+    const field = type.getFields()[fieldName]
+
+    if (!field) {
+      return undefined
+    }
+
+    let fieldType = field.type
+
+    // Unwrap NonNull and List wrappers
+    if (isNonNullType(fieldType)) {
+      fieldType = fieldType.ofType
+    }
+    if (isListType(fieldType)) {
+      fieldType = fieldType.ofType
+    }
+    if (isNonNullType(fieldType)) {
+      fieldType = fieldType.ofType
+    }
+
+    if (isObjectType(fieldType)) {
+      return fieldType.name
+    }
+
+    return undefined
+  }
+
+  /**
+   * Build invalidation events based on parsed entities
+   */
+  buildInvalidationEvents(
+    entities: EntityReference[],
+    cacheKey: string,
+    operation: "created" | "updated" | "deleted" = "updated"
+  ): InvalidationEvent[] {
+    const events: InvalidationEvent[] = []
+    const processedEntities = new Set<string>()
+
+    entities.forEach((entity) => {
+      const entityKey = `${entity.type}:${entity.id}`
+
+      if (processedEntities.has(entityKey)) {
+        return
+      }
+      processedEntities.add(entityKey)
+
+      // Get related entities for this entity type
+      const relatedEntities = entities.filter(
+        (e) => e.type !== entity.type || e.id !== entity.id
+      )
+
+      // Build cache keys that might be affected
+      const affectedKeys = this.buildAffectedCacheKeys(
+        entity,
+        relatedEntities,
+        cacheKey,
+        operation
+      )
+
+      events.push({
+        entityType: entity.type,
+        entityId: entity.id,
+        relatedEntities,
+        cacheKeys: affectedKeys,
+      })
+    })
+
+    return events
+  }
+
+  /**
+   * Build list of cache keys that should be invalidated
+   */
+  private buildAffectedCacheKeys(
+    entity: EntityReference,
+    relatedEntities: EntityReference[],
+    originalKey: string,
+    operation: "created" | "updated" | "deleted" = "updated"
+  ): string[] {
+    const keys = new Set<string>([originalKey])
+
+    // Add keys based on entity type and ID
+    keys.add(`${entity.type}:${entity.id}`)
```

Comment:
Self-review: there are some keys that we could potentially get rid off, ill think about it